### PR TITLE
(feat) Support for Custom datasources and Warnings in field validations

### PR DESCRIPTION
**Pull request by samuelmale**
#### Custom datasources
This is achieved through the `calculate.calculateExpression` at the field level. It supports any kind of arbitrary datasources as long as the data fetcher is available within the Expression runner's context. Currently the expression runner supports all API functions bundled with the engine.

```json
{
  "calculate": {
    "calculateExpression": "resolve(api.getLatestObs(patient.id, `${conceptUUID}`)).valueString"
   }
}
```

#### Warnings in field validation
These are supported by the `js_expression` validator. Eg.

```json
{
  "validators": [
    {
      "type": "js_expression",
      "warnsWhenExpression": "${expression}",
      "message": "Custom message"
    }
  ]
}
``` 

##### Linked Issues:

- https://ohri.atlassian.net/browse/OHRI-935
- https://ohri.atlassian.net/browse/OHRI-1208

### CODE REVIEW

File: src/api/api.ts [hunk 1]
```diff
@@ -43,3 +40,14 @@ export function fetchConceptNameByUuid(conceptUuid: string) {
     }
   });
 }
+
+export function getLatestObs(patientUuid: string, conceptUuid: string, encounterTypeUuid?: string) {
+  let params = `patient=${patientUuid}&code=${conceptUuid}${
```
Comment (ebambo) [inline src/api/api.ts:45]: Does this comparison compile? seems incomplete to me

File: src/components/encounter/ohri-encounter-form.tsx [hunk 2]
```diff
@@ -107,16 +92,8 @@ export const OHRIEncounterForm: React.FC<OHRIEncounterFormProps> = ({
...
-      setEncounterLocation(location);
-    }
-  }, [location]);
-
-  useEffect(() => {
-    const allFormFields: Array<OHRIFormField> = [];
-    const tempInitVals = {};
-    let isFieldEncounterBindingComplete = false;
+  const rawFormFields = useMemo(() => {
+    const ret = [];
```
Comment (ebambo) [inline src/components/encounter/ohri-encounter-form.tsx]: What is `ret` ?

File: src/components/encounter/ohri-encounter-form.tsx [hunk 3]
```diff
@@ -311,21 +265,20 @@ export const OHRIEncounterForm: React.FC<OHRIEncounterFormProps> = ({
...
-      let answerOptionid: string;
-      if (firstInvalidField.questionOptions.rendering === 'radio') {
-        answerOptionid = `${firstInvalidField.id}-${firstInvalidField.questionOptions.answers[0].label}`;
-        scrollIntoView(answerOptionid, true);
-      } else if (firstInvalidField.questionOptions.rendering === 'checkbox') {
-        answerOptionid = `${firstInvalidField.label}-input`;
-        scrollIntoView(answerOptionid, true);
-      } else {
-        scrollIntoView(firstInvalidField.id, true);
+      switch (invalidFields[0].questionOptions.rendering) {
```
Comment (ebambo) [inline src/components/encounter/ohri-encounter-form.tsx:271]: Is this change also related to adding custom datasources and warnings?

File: src/components/inputs/_input.scss [hunk 4]
```diff
@@ -149,6 +149,15 @@
   color: #da1e28;
 }
 
+.warningLabel :global(.cds--form-requirement) {
+  display: block;
+  max-height: 12.5rem;
+  overflow: visible;
+  font-weight: 400;
+  // color: #f1c21b; This is from bootstrap but Carbon defaults to black
```
Comment (ebambo) [inline src/components/inputs/_input.scss]: Do we need to keep this comment? Any further action required?

File: src/components/inputs/number/ohri-number.component.tsx [hunk 5]
```diff
@@ -17,25 +17,27 @@ const OHRINumber: React.FC<OHRIFormFieldProps> = ({ question, onChange, handler
...
   }, [question['submission']]);
 
   field.onBlur = event => {
-   if (event && event.target.value != field.value) {
-        // testing purposes only
-        field.value = event.target.value;
-        setFieldValue(question.id, event.target.value);
-      }
+    if (event && event.target.value != field.value) {
+      // testing purposes only
```
Comment (ebambo) [inline src/components/inputs/number/ohri-number.component.tsx:32]: What do you mean by testing purposes only?

File: src/components/inputs/radio/ohri-radio.component.tsx [hunk 6]
```diff
@@ -89,11 +92,14 @@ const OHRIRadio: React.FC<OHRIFormFieldProps> = ({ question, onChange, handler }
...
-            {!isFieldRequiredError && errors?.length > 0 && (
-              <div className={styles.errorLabel}>
-                <div className={`cds--form-requirement`}>{errors[0].errMessage}</div>
-              </div>
-            )}
+            {(!isFieldRequiredError && errors?.length > 0) ||
+              (warnings.length > 0 && (
+                <div className={errors.length ? styles.errorLabel : warnings.length ? styles.warningLabel : ''}>
+                  <div className={`cds--form-requirement`}>
+                    {errors.length ? errors[0].message : warnings.length ? warnings[0].message : null}
```
Comment (ebambo) [inline src/components/inputs/radio/ohri-radio.component.tsx:99]: I see this across the code `errors[0].message` | `warnings[0].message`. If we know that we will always have one element, why are we making it a list?

File: src/components/inputs/text-area/ohri-text-area.component.tsx [hunk 7]
```diff
@@ -16,10 +16,12 @@ const OHRITextArea: React.FC<OHRIFormFieldProps> = ({ question, onChange, handle
   const [errors, setErrors] = useState([]);
   const [conceptName, setConceptName] = useState('Loading...');
   const isFieldRequiredError = useMemo(() => errors[0]?.errCode == fieldRequiredErrCode, [errors]);
+  const [warnings, setWarnings] = useState([]);
 
   useEffect(() => {
-    if (question['submission']?.errors) {
-      setErrors(question['submission']?.errors);
+    if (question['submission']) {
```
Comment (ebambo) [inline src/components/inputs/text-area/ohri-text-area.component.tsx:22]: Am also seeing this `setErrors` & `setWarnings` repeated across multiple components wrapped on the `useEffect` hook. Can't this be abstracted?  Remember am not a ReactJS guru, just asking for ways to avoid code repetition.

Comment (ebambo) [reply src/components/inputs/text-area/ohri-text-area.component.tsx:22]: Am asking this also because of the number of files you needed to change just add this, I imagine what happens if you forget one or more and in case this changes in the future you will have to make this changes in all of this file again.

File: src/components/section/ohri-form-section.component.tsx [hunk 8]
```diff
@@ -8,7 +8,11 @@ import { useField } from 'formik';
 
 export const getFieldControl = (question: OHRIFormField) => {
   // Check if a concept wasn't provided
```
Comment (ebambo) [inline src/components/section/ohri-form-section.component.tsx]: Can we change the comments with clear / clean code? something like: `isConceptProvided()`

File: src/hooks/useInitialValues.ts [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+  useEffect(() => {
+    const asyncItemsKeys = Object.keys(asyncInitValues);
+    Promise.all(asyncItemsKeys.map(key => asyncInitValues[key])).then(results => {
+      asyncItemsKeys.forEach((key, index) => {
+        if (!isEmpty(results[index])) {
+          initialValues[key] = results[index];
+          const field = formFields.find(field => field.id === key);
+          try {
+            getHandler(field.type).handleFieldSubmission(field, results[index], encounterContext);
```
Comment (ebambo) [inline src/hooks/useInitialValues.ts:24]: What happens if `getHandler(field.type)` returns `null` ? I believe you are swallowing the error sending it to the console. I think we should improve this (not limited to this portion of code) in the future to show the error to the dev, something like: `Error processing filed with id: `myFieldId`, Type `myWrongFieldType` has no handler configure or is not supported`.  This kind of things would avoid a lot of errors and spare debugging time

File: src/validators/default-value-validator.ts [hunk 10]
```diff
@@ -7,18 +7,22 @@ export const OHRIDefaultFieldValueValidator: FieldValidator = {
...
       if (!field.questionOptions.answers?.find(answer => answer.concept == value)) {
-        return [{ errCode: 'invalid.defaultValue', errMessage: 'Value not found in coded answers list' }];
+        return [
+          { resultType: 'error', errCode: 'invalid.defaultValue', message: 'Value not found in coded answers list' },
+        ];
       }
     }
     if (field.questionOptions.rendering == 'date') {
       // Check if value is a valid date value
       if (!moment(value, moment.ISO_8601, true).isValid()) {
```
Comment (ebambo) [inline src/validators/default-value-validator.ts:17]: This is very restrictive, I know is not part of this PR but you should consider getting this from the client otherwise it will only work for that format.

**Review by ebambo — CHANGES_REQUESTED** [review]

File: src/api/api.ts (hunk 1 above)
Comment (samuelmale) [reply src/api/api.ts:45]: Yes it does compile

File: src/components/encounter/ohri-encounter-form.tsx (hunk 2 above)
Comment (samuelmale) [reply src/components/encounter/ohri-encounter-form.tsx]: Temporarily holds the return value. This a [common naming pattern](https://www.reddit.com/r/learnprogramming/comments/31cep5/comment/cq0a475/?utm_source=share&utm_medium=web2x&context=3) and it's in short form because `return` is a reserved keyword

File: src/components/encounter/ohri-encounter-form.tsx (hunk 3 above)
Comment (samuelmale) [reply src/components/encounter/ohri-encounter-form.tsx:271]: No, it's part of code polishing.

File: src/components/inputs/_input.scss (hunk 4 above)
Comment (samuelmale) [reply src/components/inputs/_input.scss]: There is a plan of changing Carbon's default styling for fields within a warning state to the above a color code. That's why I left it there, but I can remove it if you like.

File: src/components/inputs/radio/ohri-radio.component.tsx (hunk 6 above)
Comment (samuelmale) [reply src/components/inputs/radio/ohri-radio.component.tsx:99]: We actually support multiple warnings and errors it's just that the UX paradigm stipulates rendering one message at a go.

File: src/components/inputs/text-area/ohri-text-area.component.tsx (hunk 7 above)
Comment (samuelmale) [reply src/components/inputs/text-area/ohri-text-area.component.tsx:22]: I'm a big fan of abstraction, but we are using Functional based components which have some inherent limitations ie. No support for inheritance hierarchies. This is a local state to every control and it's managed within a local hook.

File: src/hooks/useInitialValues.ts (hunk 9 above)
Comment (samuelmale) [reply src/hooks/useInitialValues.ts:24]: I've been thinking about improving error handling within OHRI Forms and I believe it's very imperative to comprehensively support this. We need to configure error boundaries overriding the default boundary defined by the app-shell. I created a ticket for that: https://ohri.atlassian.net/browse/OHRI-1058

File: src/validators/default-value-validator.ts (hunk 10 above)
Comment (samuelmale) [reply src/validators/default-value-validator.ts:17]: The reason for the restriction is because the default value isn't defined by the client but rather the server side which uses ISO formats; But I agree we need to put the user's locale into consideration

File: src/components/inputs/number/ohri-number.component.tsx (hunk 5 above)
Comment (samuelmale) [reply src/components/inputs/number/ohri-number.component.tsx:32]: I don't clearly remember the details of issue now (it's more than a year now) but I remember having issues with Formik's `setFieldValue` not updating the value as expected within jest environments. Maybe it's something I should have another look at to figure whether it's still existing.

File: src/components/section/ohri-form-section.component.tsx (hunk 8 above)
Comment (samuelmale) [reply src/components/section/ohri-form-section.component.tsx]: Nice one, cleared.

File: src/components/encounter/ohri-encounter-form.tsx (hunk 2 above)
Comment (ebambo) [reply src/components/encounter/ohri-encounter-form.tsx]: Sorry for being old but I believe that would not be approved by uncle Bob because you never `return return;`. There is no meaning on that, `temp` or better `contextNameOfWhatYouAreReturning` is always better.  ;-)

File: src/components/inputs/_input.scss (hunk 4 above)
Comment (ebambo) [reply src/components/inputs/_input.scss]: There is non-reserved word called `TO DO`, that you can use. And I think Visual Studio Code as an extension that scans your code shows how many you have.

File: src/components/inputs/number/ohri-number.component.tsx (hunk 5 above)
Comment (ebambo) [reply src/components/inputs/number/ohri-number.component.tsx:32]: That is why you should use `TO DO` and add a description of what needs to be done or better, create a Jira Ticket.

File: src/components/inputs/text-area/ohri-text-area.component.tsx (hunk 7 above)
Comment (ebambo) [reply src/components/inputs/text-area/ohri-text-area.component.tsx:22]: Here is what you can do:

Abstract this in a hook 
```
import { useState } from 'react';

export const useValidation = () => {
  const [errors, setErrors] = useState([]);
  const [warnings, setWarnings] = useState([]);

  const addError = (error) => {
    setErrors((prevErrors) => [...prevErrors, error]);
  };

  const clearErrors = () => {
    setErrors([]);
  };

  const addWarning = (warning) => {
    setWarnings((prevWarnings) => [...prevWarnings, warning]);
  };

  const clearWarnings = () => {
    setWarnings([]);
  };

  return {
    errors,
    warnings,
    addError,
    clearErrors,
    addWarning,
    clearWarnings,
  };
};
```

And take advantage of the attraction hooks can provide.

```
import { useValidation } from './useValidation';

const MyComponent = () => {
  const { addError, addWarning } = useValidation();

  useEffect(() => {
    // Do some validation
    if (/* validation fails */) {
      addError('Validation failed');
    } else if (/* validation produces warning */) {
      addWarning('Validation produced warning');
    }
  }, []);

  // ...
};
```
I might be wrong thou

File: src/hooks/useInitialValues.ts (hunk 9 above)
Comment (ebambo) [reply src/hooks/useInitialValues.ts:24]: Thanks

File: src/components/inputs/text-area/ohri-text-area.component.tsx (hunk 7 above)
Comment (samuelmale) [reply src/components/inputs/text-area/ohri-text-area.component.tsx:22]: At the end of the day you are controlling the `useValidation()` hook through a local `useEffect` to the component. I love to keep things simple. Abstraction isn't always a perfect design. Sometimes it introduces complexity within the app. Have a look at this article: https://www.deconstructconf.com/2019/dan-abramov-the-wet-codebase

File: src/components/inputs/number/ohri-number.component.tsx (hunk 5 above)
Comment (samuelmale) [reply src/components/inputs/number/ohri-number.component.tsx:32]: I have spiked on this issue again, commented out the line 33 and all the tests relating to the number input are failing; let's ignore this for now untill we figure a better fix for the tests.

File: src/components/inputs/_input.scss (hunk 4 above)
Comment (samuelmale) [reply src/components/inputs/_input.scss]: Removed

File: src/components/encounter/ohri-encounter-form.tsx (hunk 2 above)
Comment (samuelmale) [reply src/components/encounter/ohri-encounter-form.tsx]: Polished

File: src/components/inputs/number/ohri-number.component.tsx (hunk 5 above)
Comment (samuelmale) [reply src/components/inputs/number/ohri-number.component.tsx:32]: Actually the issue is to do with how Carbon propagates events and harvests values out of a blur which is buggy within the Jest env

**Review by ebambo — APPROVED** [review]

**Comment by ebambo** [comment]
Next time only push things related to the task, this PR was too long ufff BUT LGTM
